modules/seqexec/server/src/main/scala/seqexec/server/gmos/Gmos.scala

```diff
@@ -54,17 +57,47 @@ abstract class Gmos[F[_]: MonadError[?[_], Throwable]: Concurrent: Logger, T <:
 
   override val keywordsClient: KeywordsClient[F] = this
 
-  val continueCommand: Time => F[ObserveCommandResult] =
-    controller.resumePaused
+  val nsCmdRef: Ref[F, Option[NSObserveCommand]] = nsCmdR
+
+  val nsCount: F[Int] = controller.nsCount
 
-  override val observeControl: InstrumentSystem.CompleteControl[F] = CompleteControl(
-    StopObserveCmd(controller.stopObserve),
-    AbortObserveCmd(controller.abortObserve),
-    PauseObserveCmd(controller.pauseObserve),
-    ContinuePausedCmd(continueCommand),
-    StopPausedCmd(controller.stopPaused),
-    AbortPausedCmd(controller.abortPaused)
-  )
+  override def observeControl(config: CleanConfig): InstrumentSystem.CompleteControl[F] =
+    if(isNodAndShuffle(config))
+      CompleteControl(
+        StopObserveCmd(stopNS),
+        AbortObserveCmd(abortNS),
+        PauseObserveCmd(pauseNS),
+        ContinuePausedCmd(controller.resumePaused),
+        StopPausedCmd(controller.stopPaused),
+        AbortPausedCmd(controller.abortPaused)
+      )
+    else
+      CompleteControl(
+        StopObserveCmd(_ => controller.stopObserve),
+        AbortObserveCmd(_ => controller.abortObserve),
+        PauseObserveCmd(_ => controller.pauseObserve),
+        ContinuePausedCmd(controller.resumePaused),
+        StopPausedCmd(controller.stopPaused),
+        AbortPausedCmd(controller.abortPaused)
+      )
+
+  def stopNS(gracefully: Boolean): F[Unit] =
+    if(gracefully)
+      nsCmdRef.set(NSObserveCommand.StopGracefully.some)
```

Comment:
I'm not totally familiar with `Ref` but I think it is better to use `update` rather than `set` to ensure updates are safe in presence of concurrence

Reply:
But the code documentation fro `update` says:
```
   * Satisfies:
   *   `r.update(_ => a) == r.set(a)`
```

Reply:
You can check Flabio's talk about this. As I understand it, it may happen that updates are lost if multiple threads call set

https://vimeo.com/366191463

TBH I don't know if in practice it makes any difference for us